Implemented Finite State Machine

gregormaclaine opened the pull request:
This feature branch creates a FSM framework for the program that will make it much easier to have different views that can be switched between such as the menu and the game. This also involved clearing up the code for the menu and modularising some of the features in it like the button.

The most recent commits in main have been merged into it, meaning it includes the new additions of player health and the background.

This version is also fully functional on macOS amd64, so it needs testing on windows before mergining into the main branch.

gregormaclaine committed:
Added sprite class and new enemy image

gregormaclaine committed:
Implemented sprite class in player and staticsprite

gregormaclaine committed:
Added third platform

gregormaclaine committed:
Added delta time fix

This commit technically does not work because for some reason at the start of the program the framerate acts a bit weird.
It causes everything effected by gravity (player and enemies) to fall directly through the platforms and therefore go beneath the screen.
This then immediately fixes itself as the framerate stabalises.

I am not going to try to fix this because when there is a menu beforehand, it will allow time before the players and enemies are placed, meaning that issue should never occur.

gregormaclaine committed:
Merge branch 'main' into sprite-class-again

# Conflicts:
#	src/main.cpp

gregormaclaine committed:
Added control-r to also reset enemies positions

gregormaclaine committed:
Merge branch 'main' into sprite-class

# Conflicts:
#	src/main.cpp

gregormaclaine committed:
TEMP: states

gregormaclaine committed:
Edits for Mac installation

gregormaclaine committed:
Added update function to sprite class

gregormaclaine committed:
Updates to vscode settings

gregormaclaine committed:
TMP: Added main functionality of menu buttons and finite state machine
This is a temporary commit, as currently the user stays in the menu stage and can't progress to the game due to the buttons not being fully functional.

gregormaclaine committed:
Moved game logic to Game class and finished FSM

gregormaclaine committed:
Merge branch 'main' into finite-state-machine

# Conflicts:
#	src/Enemy.cpp
#	src/Enemy.hpp
#	src/Menu.cpp
#	src/Menu.hpp
#	src/Player.cpp
#	src/Player.hpp
#	src/main.cpp

egorvert reviewed (changes requested):
Code is good and functional on Mac. It would be appreciated if you could add comments to your code to explain what it does so the rest of the team doesn't have to spend their precious time on figuring out what the hell you're doing.

Some bugs I noticed:
- Hitboxes are always shown and control / command B does not turn them off
- Resizing a window causes the character to fly off the screen

Tl;Dr
★★★★☆
Good feature rich commit but with little annotation and a few bugs. Also needs testing on Windows.

gregormaclaine commented:
> Code is good and functional on Mac. It would be appreciated if you could add comments to your code to explain what it does so the rest of the team doesn't have to spend their precious time on figuring out what the hell you're doing.
> 
> Some bugs I noticed:
> 
> * Hitboxes are always shown and control / command B does not turn them off
> * Resizing a window causes the character to fly off the screen
> 
> Tl;Dr ★★★★☆ Good feature rich commit but with little annotation and a few bugs. Also needs testing on Windows.

gregormaclaine commented:
Woops closed it by accident. Also I see the bug with resizing the window, but the hitbox outlines work perfectly find for me? They don't show up to begin with and toggle visibility with Ctrl-B.

gregormaclaine committed:
Added a lot of comments

gregormaclaine committed:
Added some docs on the FSM

Commit:
Fixed bugs to work on windows

Commit:
Fixed hitbox always showing bug on windows

gregormaclaine committed:
Sprites now don't clip through platforms during window resize

gregormaclaine committed:
Separated deltatime checker into separate file

gregormaclaine committed:
Increased certainty of deltatime anomaly & fixed enemies clipping through ground on gamestart

gregormaclaine commented:
All bugs and changes requested by Egor have been implemented.
Branch is now ready for merge

gregormaclaine committed:
Hopefully fixed warning in vscode settings from deprecated key

gregormaclaine committed:
Changed enemies and platforms to std::vector from basic C array

std::vector is dynamic so it can be freely added to, which will be important later on.
It's also much easier to iterate through.

I needed the vectors to be of the class pointers as sometimes (by that I mean all the time) the textures of the sprites would unload since the vector would keep trying to move the memory around which would break the reference between the texture and the sprite causing it to show a tan rectangle instead of the actual picture.
